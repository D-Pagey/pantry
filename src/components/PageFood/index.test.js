import React from 'react';
import { Redirect, useParams } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { CategoriesArray, Fridge } from '../../fixtures';
import { PageFood } from '.';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(() => ({
        category: 'meat'
    })),
    Redirect: jest.fn(() => null),
    useHistory: () => ({
        push: mockHistoryPush
    })
}));

const context = {
    categories: CategoriesArray,
    fridge: Fridge,
    deleteFoodItem: () => {}
};

describe('PageFood component', () => {
    it('should render', () => {
        const { container } = render(<PageFood />, context);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a loading spinner initially', () => {
        const overrideContext = {
            categories: [],
            fridge: []
        };

        const { getByTestId } = render(<PageFood />, { ...context, ...overrideContext });
        getByTestId('loading');
    });

    it('when the category is all, it should render all food', () => {
        useParams.mockImplementation(() => ({
            category: 'all'
        }));

        const { getByText } = render(<PageFood />, context);
        Fridge.map((item) => getByText(item.name));
    });

    it('when the category is not all, it should filter down fridge', () => {
        useParams.mockImplementation(() => ({
            category: CategoriesArray[1].name
        }));

        const { getByText, queryByText } = render(<PageFood />, context);

        Fridge.map((item) => {
            if (item.categories.includes(CategoriesArray[1].id)) {
                return getByText(item.name);
            }

            return expect(queryByText(item.name)).toBe(null);
        });
    });

    it('when the category doesnt exist, it should redirect', () => {
        useParams.mockImplementationOnce(() => ({
            category: 'hello'
        }));

        render(<PageFood />, context);
        expect(Redirect).toHaveBeenCalledWith({ to: '/not-found' }, expect.any(Object));
    });

    it('should render a message when there is no data for a category', () => {
        useParams.mockImplementation(() => ({
            category: CategoriesArray[3].name
        }));

        const { getByTestId } = render(<PageFood />, context);
        getByTestId('pageFoodNoData');
    });

    it('should handle an edit food click', () => {
        useParams.mockImplementation(() => ({
            category: 'all'
        }));

        const updateFridge = jest.fn();
        const { getAllByTestId } = render(<PageFood />, { ...context, updateFridge });
        const button = getAllByTestId('editButton')[0];

        userEvent.click(button);

        expect(mockHistoryPush).toHaveBeenCalledWith('/add', Fridge[0]);
    });

    it.todo('should refilter if category changes');
});
