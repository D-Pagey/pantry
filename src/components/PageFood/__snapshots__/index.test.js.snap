// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageFood component should render 1`] = `
.c1 {
  color: red;
  list-style: none;
}

.c0 {
  cursor: pointer;
}

.c2 {
  background-color: #288efc;
  border: 0;
  border-radius: 0.25rem;
  color: white;
  cursor: pointer;
  font-size: 1rem;
  height: 45px;
  line-height: 1.2;
  width: 200px;
}

<div>
  <h1>
    meat
  </h1>
  <div>
    <table>
      <thead>
        <tr>
          <th
            class="c0"
          >
            Name
          </th>
          <th
            class="c0"
            data-testid="foodTableCategoryColumn"
          >
            Categories
          </th>
          <th
            class="c0"
          >
            Expires
          </th>
          <th
            class="c0"
          >
            Servings
          </th>
          <th
            class="c0"
          >
            Amend
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            data-testid="foodTableName"
          >
            chicken
          </td>
          <td>
            meat
          </td>
          <td>
            <li
              class="c1"
            >
              12th Oct
            </li>
          </td>
          <td>
            2
          </td>
          <td>
            <button
              data-testid="editButton"
              style="cursor: pointer; margin: 0px 1rem 0px 0px;"
              type="button"
            >
              <img
                alt="edit"
                src="edit.svg"
              />
            </button>
            <button
              data-testid="deleteButton"
              style="cursor: pointer;"
              type="button"
            >
              <img
                alt="delete"
                src="delete.svg"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td
            data-testid="foodTableName"
          >
            steak
          </td>
          <td>
            meat
          </td>
          <td>
            <li
              class="c1"
            >
              9th Apr
            </li>
          </td>
          <td>
            1
          </td>
          <td>
            <button
              data-testid="editButton"
              style="cursor: pointer; margin: 0px 1rem 0px 0px;"
              type="button"
            >
              <img
                alt="edit"
                src="edit.svg"
              />
            </button>
            <button
              data-testid="deleteButton"
              style="cursor: pointer;"
              type="button"
            >
              <img
                alt="delete"
                src="delete.svg"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <a
    href="/add"
  >
    <button
      class="c2"
      data-testid=""
      type="button"
      width=""
    >
      Add Item
    </button>
  </a>
</div>
`;
